### PR TITLE
Kat Eastman - Munker - Octos

### PR DESCRIPTION
**Pull request by kseastman**
# API Muncher
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
How did you go about exploring the Edamam API, how did you try querying the API? | I read through the provided API instructions, and then tested queries in PostMan.
Describe your API Wrapper.  How did you decide on the methods you created? |  Initially I had one large search method, that I broke into three methods overall, the original search method called in my code and two helper methods. It submits a query using a base url, and then takes the response and sends it through a factory method in a recipe class.
Describe an edge case or failure case test you wrote for your API Wrapper. | What happens if a recipe search contains an invalid string, what happens if a keyword search contains invalid characters?
Explain how VCR aids in testing an API. | VCR allows for a single call to the API to be saved as a cassette, so subsequent testing can use that saved version, rather than sending out additional calls. This lets tests happen when there is no direct internet connection to the API, reduces the costs in both time and api limits for running a test and development server.
What is the Heroku URL of your deployed application? | http://kse-muncher.herokuapp.com/
Provide a link to the Trello board you used | https://trello.com/b/lPQGiknu/ke-api-muncher

### CODE REVIEW

File: config/routes.rb [hunk 1]
```diff
@@ -0,0 +1,13 @@
+Rails.application.routes.draw do
+get '/', to: "recipes#root", as: :root
+get '/:query/results', to: "recipes#results", as: :results
+get '/:id', to: 'recipes#show', as: :show
```
Comment (droberts-sea) [inline config/routes.rb:4]: Your routes here aren't restful! I would probably use `resources :recipes, only: [:index, :show]` for the search results and show pages, with the search term as a query param instead of part of the URL.

File: lib/edamam_api_wrapper.rb [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+  def self.keyword_search(query)
+    url = BASE_URL + "q=#{query}" "&app_id=#{APP}" + "&app_key=#{KEY}" + "&from=0&to=1000"
+    url = URI.encode(url)
+    data = HTTParty.get(url)
+
+    if !raise_on_error(data)
+      return data["hits"].map do |raw_recipe|
+        recipe = raw_recipe["recipe"]
+        Recipe.from_api(recipe)
```
Comment (droberts-sea) [inline lib/edamam_api_wrapper.rb:41]: Since all `raise_on_error` does is conditionally throw an exception, I'm not sure it makes sense to check the return value.

File: lib/edamam_api_wrapper.rb [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+      end
+    else
+    end
+
+  end
+
+  private
+  def self.raise_on_error(response)
+    unless response["more"]
+      raise EdamamError.new("There were no results for #{response["q"]}")
```
Comment (droberts-sea) [inline lib/edamam_api_wrapper.rb:51]: I don't think that checking the `more` key will work for the single recipe search, since it sends back an array.

File: test/controllers/recipes_controller_test.rb [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+      get show_path(slider_uri)
+
+      must_respond_with :success
+    end
+  end
+
+  it "handles an invalid request" do
+    VCR.use_cassette('Slider') do
+      slider_uri = "#recipe_slider"
+      proc { get show_path(slider_uri)}.must_raise
```
Comment (droberts-sea) [inline test/controllers/recipes_controller_test.rb:24]: I like that you've identified this as a test case - good work! However, I'm not sure I agree with your engineering decision about how the app ought to behave in this case. In general, controllers shouldn't ever expose exceptions to the user, but should send back an appropriate status code and let the user know what happened via flash.

**Comment by droberts-sea** [comment]
# API Muncher
## What We're Looking For

Feature | Feedback
:------------- | :-------------
**Core Requirements** | 
Git hygiene | yes
Comprehension questions | yes
**General** | 
Rails fundamentals (RESTful routing, use of named paths) | see inline
Semantic HTML | yes
Errors are reported to the user | works for search, but when I mangle the recipe URI, I get an exception rather than a polite error message
API Wrapper to handle the API requests | yes
Controller testing | yes
Lib testing | yes
Search Functionality | yes
List Functionality | yes
Show individual item functionality (link to original recipe opens in new tab) | yes
**Styling** | 
Responsive layout | yes
List View shows 10 items at a time/pagination | yes
The app is styled to create an attractive user interface | yes
**API Features** | 
The App attributes Edaman | yes
The VCR casettes do not contain the API key | yes
**External Resources** | 
Link to Trello Board | yes
Link to deployed app on Heroku | yes
**Overall** | Great work overall! Your code is well-organized and easy to read, and it's clear the learning goals for this assignment were met. Theres some room for improvement around your error handling - making sure there's a comprehensive story from bottom to top about how errors are spotted and reported to the user is one of the things that separates a good app from a great one. Other than that I'm happy with this submission - I especially like how solid your test coverage is. Keep up the hard work!
